src/framework/audio/internal/platform/lin/pwaudiodriver.cpp

```diff
@@ -0,0 +1,807 @@
+        m_devicesChanged.notify();
+    }
+}
+
+// Called from "pw-driver-loop" thread
+void PwRegistry::c_register_global(void* data, uint32_t id, uint32_t /* permissions*/, const char* type, uint32_t /*version*/,
+                                   const struct spa_dict* props)
+{
+    auto registry = static_cast<PwRegistry*>(data);
+    registry->registerGlobal(id, type, props);
+}
+
+// Called from "pw-driver-loop" thread
+void PwRegistry::registerGlobal(uint32_t id, const char* type, const struct spa_dict* props)
+{
+    // For playback, we are interested in Audio/Sink only.
+    // We fetch also Audio/Device, because they arguably have neater nick & description than Audio/Sink.
+    // See `getDevices` for details.
+
+    if (spa_streq(type, PW_TYPE_INTERFACE_Device)) {
+        auto mediaClass = spa_dict_lookup(props, PW_KEY_MEDIA_CLASS);
+        if (spa_streq(mediaClass, "Audio/Device")) {
+            auto name = spa_dict_lookup(props, PW_KEY_DEVICE_NAME);
+            auto nick = spa_dict_lookup(props, PW_KEY_DEVICE_NICK);
+            auto desc = spa_dict_lookup(props, PW_KEY_DEVICE_DESCRIPTION);
+
+            std::scoped_lock<std::mutex> lock(m_mutex);
+            m_pwDevices.push_back(Device {
+                id,
+                name,
+                nick ? nick : "",
+                desc ? desc : ""
+            });
+        }
+    } else if (spa_streq(type, PW_TYPE_INTERFACE_Node)) {
+        auto mediaClass = spa_dict_lookup(props, PW_KEY_MEDIA_CLASS);
+        if (spa_streq(mediaClass, "Audio/Sink")) {
+            auto name = spa_dict_lookup(props, PW_KEY_NODE_NAME);
+            auto nick = spa_dict_lookup(props, PW_KEY_NODE_NICK);
+            auto desc = spa_dict_lookup(props, PW_KEY_NODE_DESCRIPTION);
+            auto deviceIdStr = spa_dict_lookup(props, PW_KEY_DEVICE_ID);
+            auto deviceId = static_cast<uint32_t>(deviceIdStr ? std::strtoul(deviceIdStr, nullptr, 10) : 0);
+
+            LOGI() << "Registering device: " << name << " (" << deviceId << ")";
+
+            std::scoped_lock<std::mutex> lock(m_mutex);
+            m_pwNodes.push_back(SinkNode {
+                id,
+                name,
+                nick ? nick : "",
+                desc ? desc : "",
+                deviceId
+            });
+            _devicesDirty();
+        }
+    }
+}
+
+// Called from "pw-driver-loop" thread
+void PwRegistry::c_unregister_global(void* data, uint32_t id)
+{
+    auto registry = static_cast<PwRegistry*>(data);
+    registry->unregisterGlobal(id);
+}
+
+// Called from "pw-driver-loop" thread
+void PwRegistry::unregisterGlobal(uint32_t id)
+{
+    std::scoped_lock<std::mutex> lock(m_mutex);
+
+    const auto numN = m_pwNodes.size();
+
+    m_pwDevices.erase(std::remove_if(m_pwDevices.begin(), m_pwDevices.end(), [id](const Device& device) {
+        return device.id == id;
+    }), m_pwDevices.end());
+
+    m_pwNodes.erase(std::remove_if(m_pwNodes.begin(), m_pwNodes.end(), [id](const SinkNode& node) {
```

Comment:
You can use muse::remove_if (from containers.h). It's much more convenient to use